SAT/Z3Interfacing.cpp

```diff
@@ -294,46 +320,46 @@ SATSolver::Status Z3Interfacing::solveUnderAssumptions(const SATLiteralStack& as
   return result;
 }
 
-SATSolver::VarAssignment Z3Interfacing::getAssignment(unsigned var) 
+SATSolver::VarAssignment Z3Interfacing::getAssignment(unsigned var)
 {
   CALL("Z3Interfacing::getAssignment");
   BYPASSING_ALLOCATOR;
 
   ASS_EQ(_status,SATISFIABLE);
   bool named = isNamedExpr(var);
   z3::expr rep = named ? getNameExpr(var) : getRepresentation(SATLiteral(var,1)).expr;
+  outputln("(get-value (", rep, "))");
```

Comment:
Does the z3  documentation say that get-value and model.eval do the same thing? It would be interesting to check. We've had issues in the past with eval being a bit lazy... and the model completion option (not there for get-value) is necessary.

Reply:
No it doesn't. I just used that as I thought it would be the intuitive equivalent. 
This is what the documentation says:
```

/**
   \brief Evaluate the AST node \c t in the given model.
   Return \c true if succeeded, and store the result in \c v.

   If \c model_completion is \c true, then Z3 will assign an interpretation for any constant or function that does
   not have an interpretation in \c m. These constants and functions were essentially don't cares.

   If \c model_completion is \c false, then Z3 will not assign interpretations to constants for functions that do
   not have interpretations in \c m. Evaluation behaves as the identify function in this case.

   The evaluation may fail for the following reasons:

   - \c t contains a quantifier.

   - the model \c m is partial, that is, it doesn't have a complete interpretation for uninterpreted functions.
   That is, the option \ccode{MODEL_PARTIAL=true} was used.

   - \c t is type incorrect.

   - \c Z3_interrupt was invoked during evaluation.

   def_API('Z3_model_eval', BOOL, (_in(CONTEXT), _in(MODEL), _in(AST), _in(BOOL), _out(AST)))
*/
Z3_bool_opt Z3_API Z3_model_eval(Z3_context c, Z3_model m, Z3_ast t, bool model_completion, Z3_ast * v);
```
But I think there might be a discrebancy between `evaluate` and `get-value` indeed. Or at least this would explain an issue I just ran into while looking for a bug.

Reply:
Okay I dug into the z3 code, and found out that there is only one difference between `get-value` and `model.evaluate(...)`. That is, in the case of `get-value`, the model is being compressed before it is evaluated. 
So I think we can keep it as is. 
This also explains the discrepancy I ran into.